examples/tensorflow/text-classification/run_text_classification.py

```diff
@@ -452,85 +414,129 @@ def preprocess_function(examples):
-        # If there's a validation dataset but no training set, just evaluate the metrics
-        eval_dataset = DataSequence(
-            eval_dataset, non_label_column_names, batch_size=training_args.per_device_eval_batch_size, labels=True
+        # endregion
+
+        # region Optimizer, loss and compilation
+        optimizer = tf.keras.optimizers.Adam(
+            learning_rate=training_args.learning_rate,
+            beta_1=training_args.adam_beta1,
+            beta_2=training_args.adam_beta2,
+            epsilon=training_args.adam_epsilon,
+            clipnorm=training_args.max_grad_norm,
         )
-        logger.info("Computing metrics on validation data...")
         if is_regression:
-            loss = model.evaluate(eval_dataset)
-            logger.info(f"Loss: {loss:.5f}")
+            loss = tf.keras.losses.MeanSquaredError()
+            metrics = []
         else:
-            loss, accuracy = model.evaluate(eval_dataset)
-            logger.info(f"Loss: {loss:.5f}, Accuracy: {accuracy * 100:.4f}%")
-    # endregion
-
-    # region Prediction
-    if "test" in datasets:
-        logger.info("Doing predictions on test dataset...")
-
-        test_dataset = DataSequence(
-            test_dataset, non_label_column_names, batch_size=training_args.per_device_eval_batch_size, labels=False
-        )
-        predictions = model.predict(test_dataset)["logits"]
-        predictions = np.squeeze(predictions) if is_regression else np.argmax(predictions, axis=1)
-        output_test_file = os.path.join(training_args.output_dir, "test_results.txt")
-        with open(output_test_file, "w") as writer:
-            writer.write("index\tprediction\n")
-            for index, item in enumerate(predictions):
-                if is_regression:
-                    writer.write(f"{index}\t{item:3.3f}\n")
-                else:
-                    item = model.config.id2label[item]
-                    writer.write(f"{index}\t{item}\n")
-        logger.info(f"Wrote predictions to {output_test_file}!")
-    # endregion
+            loss = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+            metrics = ["accuracy"]
+        model.compile(optimizer=optimizer, loss=loss, metrics=metrics)
+        # endregion
+
+        # region Convert data to TF format
+
+        # Convert data to a tf.keras.utils.Sequence object for training if we're not using a TPU
+        # For TPU, convert to a tf.data.Dataset
+        tf_data = dict()
+        max_samples = {
+            "train": data_args.max_train_samples,
+            "validation": data_args.max_val_samples,
+            "test": data_args.max_test_samples,
+        }
+        for key in ("train", "validation", "test"):
+            if key not in datasets:
+                tf_data[key] = None
+                continue
+            if key in ("train", "validation"):
+                use_labels = True
+            else:
+                use_labels = False
```

Comment:
We could have a test set with labels (people want to evaluate on the test set of GLUE for instance) so the test should be if there is a `"labels"` columns IMO.

Reply:
This is a good point - right now the code is really treating the 'test' set as the set the user wants predictions on, and uses 'validation' as the set to compute accuracy and metrics on. This is probably not very clear. Would it make more sense to have a separate --prediction_inputs file or something like that?

Reply:
I would align to the PyTorch script that returns:
- the predictions if no labels
- the predictions + metrics if labels

Reply:
Done!